yottadb: init at r2.02, nixos/yottadb: init

ztmr opened the pull request:
###### Description of changes

This is to add YottaDB core engine, more info on: https://yottadb.com/ and https://gitlab.com/YottaDB.
There may be more packages in the future: Octo, AIM, various plugins, language bindings, etc.

###### Things done

Introducing a new package and a module required for YottaDB's correct operation.

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux (not explicitly tested yet; the code should be supported on this platform however)
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

I have tested basic functionality, and I am currently using this package together with YDBRust language wrapper (https://gitlab.com/YottaDB/Lang/YDBRust) without any issues.

This is my first contribution to nixpkgs, so please let me know if you require any more work or changes. I've followed the contribution guide in terms of commit msgs, etc. but might have missed something.

Commit:
yottadb: init at version r1.34

Add YottaDB, a proven Multi-Language NoSQL database engine.

Commit:
nixos/yottadb: init at version r1.34

Add a new YottaDB module to SUID-wrap `gtmsecshr'
executable. This is required by YDB to operate correctly.

More details to be found in YottaDB's Admin and Ops Guide,
Appendix C, YottaDB Security Philosophy, gtmsecshr Exception:
  https://docs.yottadb.com/AdminOpsGuide/securityph.html#id5

YottaDB will require `programs.yottadb.enable` to be `true`
for most of its functionality.

ztmr commented:
Had to correct the module. It's been tested in a standalone ISO/QEMU environment just now and all working as expected.

Not doing any more changes on this PR until advised otherwise.

ajs124 commented on pkgs/development/yottadb/default.nix:
```diff
@@ -0,0 +1,168 @@
+    homepage = "https://gitlab.com/YottaDB/DB/YDB";
+    license = licenses.agpl3;
+    maintainers = [ maintainers.ztmr ];
+    platforms = [ "aarch64-linux" "armv7l-linux" "armv6l-linux" "x86_64-linux" ];
```
this can probably just be `platforms.linux`

ztmr replied:
I am not so sure about this. YottaDB is a specific piece of software -- it has some parts written in assembler for its target platform and was once supported on various platforms like VAX, Alpha, SPARC, HP9000, and x86, x86_64 were added later, and ARM-flavours were added only recently. There is also RISC-V merge request that might be merged eventually...
So I think it makes sense to list these supported platforms explicitly...

ztmr commented on pkgs/development/yottadb/default.nix:
```diff
@@ -0,0 +1,168 @@
+    owner = "YottaDB";
+    repo = "DB/YDB";
+    rev = "${version}";
+    leaveDotGit = true;
```
It does not -- git metadata is used to generate version information at build time.

ztmr commented on pkgs/development/yottadb/default.nix:
```diff
@@ -0,0 +1,168 @@
+  enableParallelBuilding = true;
+
+  nativeBuildInputs = with pkgs; [
+    file cmake gnumake curl tcsh
```
OK, that is fine, but what if `stdenv.mkDerivation` was suddenly refactored and this automatic dependency was gone? Is it not better to specify all dependencies for each package explicitly, so it does not make any assumptions on anything else? If the dependency is already satisfied, there is no harm to it anyway, right?

ajs124 replied:
Maybe it would have been better to specify all dependencies for each package explicitly, but that's not how it's done right now and this won't change anytime soon, because it would basically mean having to change every single nix package.

The harm of adding random dependencies is that this might actually somehow break the build, e.g. when cross compiling plus it just adds code that, at best, does nothing

ztmr commented on pkgs/development/yottadb/default.nix:
```diff
@@ -0,0 +1,168 @@
+{
+  lib, gcc10Stdenv, fetchFromGitLab, pkgs, writeText
```
I will change the `pkgs`, that is a good idea to do.
The software can be built using GCC or Clang, I will do some more experiments to this. Is there any specific recommendation you have? I think I've just copied this `gcc10Stdenv` from some other package that compiles a C code.

ztmr commented on pkgs/development/yottadb/default.nix:
```diff
@@ -0,0 +1,168 @@
+
+gcc10Stdenv.mkDerivation rec {
+  pname = "yottadb";
+  version = "r1.34";
```
No problem, will change, makes sense.

ztmr commented on pkgs/development/yottadb/default.nix:
```diff
@@ -0,0 +1,168 @@
+    #BUILD_TYPE=Debug
+
+    mkdir build && cd build
+    cmake \
```
Should be, I will change this.

Is there any mechanism to build debug flavour of packages? I think it is a good idea to have production by default, but have an option to install debug version as well.

ajs124 commented on pkgs/development/yottadb/default.nix:
```diff
@@ -0,0 +1,198 @@
+  # and libdir=@@YDB_LIBDIR@@ breaks dependent software (YDBRust for
+  # instance) for some reason.
+  yottadbPkgConfigTemplate = writeText "yottadb.pc.in" ''
+prefix=@@YDB_DIST@@
```
This can be indented normally, using spaces, because the string builder removes the spaces. I think. Maybe verify that…

Commit:
yottadb: add a basic test script

ztmr commented on pkgs/development/yottadb/default.nix:
```diff
@@ -0,0 +1,192 @@
+    # We cannot set SUID for gtmsecshr, it has to be done using Nix Security Wrapper
+    cd $out/dist
+    mv gtmsecshr gtmsecshr.nosuid
+    ln -rfs /run/wrappers/bin/gtmsecshr .
```
@ajs124 is there any better way of doing this? This simply assumes that the wrapper is there (module enabled).

I can only think of creating yet another wrapper (to replace this symlink) that would detect if the SUID wrapper is available and either call it or fail. Is there any other way?

Commit:
yottadb: nixos-specific fixes

YottaDB does not work correctly on NixOS out of the box
in multiuser environments and requires a special SUID
binary to allow processes to communicate with each other.

This commit is trying to patch YDB to make this happen.

These changes, however, are meant to enable people to use
YottaDB for development on NixOS, but should not be used
in production.

ztmr commented:
Just to update on this, the current problem with this package is that [YottaDB relies on SUID helper](https://docs.yottadb.com/AdminOpsGuide/securityph.html#gtmsecshr-exception) that has to be in a path relative to the main binary and with certain permissions. This is checked at runtime and if this helper lacks the expected permissions or is a symlink to security wrapper, or some other attempt to hack around this, YDB won't work properly.

Nix store simply does not have flexibility to host this kind of application. There are two workarounds I can think of:
- having a NixOS specific patch to modify the YDB code to expect the SUID helper in a specific path, but there is a lot of other issues with that -- it operates in its private directory, so only security wrapper is not sufficient.
- having a wrapper script that will use Nix store only as a bootstrap and "install" YDB to a user's private (configurable) dir with all the correct permissions. This is not ideal but has some potential to work. However, normal users cannot set SUID on the helper binary :-(

Is this something that can be achieved with `systemd-tmpfiles.d` syncing the YDB runtime directory away from Nix store and setting the right permissions on everything?

Any other ideas? Who would be the best person to consult on this?

Relevant issues in YDB project, just for reference:
- https://gitlab.com/YottaDB/DB/YDB/-/issues/862
- https://gitlab.com/YottaDB/DB/YDB/-/issues/491#note_895014698

Relevant code:
- https://gitlab.com/YottaDB/DB/YDB/-/blob/r1.36/sr_unix/gtmsecshr.c
- https://gitlab.com/YottaDB/DB/YDB/-/blob/r1.36/sr_unix/gtmsecshr_wrapper.c

Commit:
yottadb: update r1.34 to r1.36

Commit:
yottadb: update to r1.38

Commit:
yottadb: use libxcrypt

Commit:
yottadb: move away from the (broken) gtmsecshr wrappers

Commit:
maintainers: fix sorting and structure post-rebase

ztmr commented:
@ajs124 not sure why the checksum mismatch is happening on `aarch64`. Any clues?

AndersonTorres commented:
> @ajs124 not sure why the checksum mismatch is happening on `aarch64`. Any clues?

I have bumped on this case a long time ago.

In a Linux OS the filesystems are case-sensitive by default, however in MacOS the filesystems are case-INsensitive by default.

Sometimes the tarball you unpacked has some files with mixed cases. They are read literally by Linux but are modified by MacOS.

Because of it, the filelists generated by both are "different", generating different checksums.

I suggest you to mark it as broken on Darwin.

P.S.: are you still there?

AndersonTorres commented on maintainers/maintainer-list.nix:
```diff
@@ -18858,6 +18858,15 @@
       fingerprint = "7AFB C595 0D3A 77BD B00F  947B 229E 63AE 5644 A96D";
     }];
   };
+  ztmr = {
```
**Split**

ztmr replied:
What do you want me to split?

AndersonTorres commented on nixos/modules/programs/yottadb.nix:
```diff
@@ -0,0 +1,40 @@
+{ config, pkgs, lib, ... }:
+
+with lib;
```
```suggestion
```

ztmr replied:
removed

AndersonTorres commented on pkgs/development/yottadb/default.nix:
```diff
@@ -0,0 +1,225 @@
+    ./01-gtmsecshr-relocate.patch
+  ];
+
+  NIX_CFLAGS_COMPILE = "-DYDB_EXTERNAL_SECSHR_PARENT_DIR=\"${ydbSecRunDir}\"";
```
```suggestion
  env.NIX_CFLAGS_COMPILE = "-DYDB_EXTERNAL_SECSHR_PARENT_DIR=\"${ydbSecRunDir}\"";
```

AndersonTorres commented on pkgs/development/yottadb/default.nix:
```diff
@@ -0,0 +1,225 @@
+
+  enableParallelBuilding = true;
+
+  nativeBuildInputs = buildInputs ++ [
```
```suggestion
  nativeBuildInputs = [
```

1. Be explicit.
2. Why??

ztmr replied:
OK, it is explicit now. There is a big overlap between runtime and build time deps because YottaDB is a compiler, and it compiles itself and various library code at build time, but also compiles things at runtime during its normal production operation. I can revisit the dependencies, but it is a bit messy to keep the list duplicated.

AndersonTorres commented on pkgs/development/yottadb/default.nix:
```diff
@@ -0,0 +1,225 @@
+
+  prePatch = ''
+    # Revert any potential repo modifications to keep the release clean
+    git reset --hard
```
Why are you manipulating the git repo?

ztmr replied:
The repo does not seem to be clean (is Nix manipulating the checked out workspace?) which breaks the mechanisms of generating the version strings baked into the build process and resulting binaries. Related to the suggested replacement of `git-describe` with a fixed string.

ztmr replied:
Removed, injecting the git ref directly into the build by manipulating the upstream cmake recipes not to use git.

AndersonTorres commented on pkgs/development/yottadb/default.nix:
```diff
@@ -0,0 +1,225 @@
+  preConfigure = ''
+    # The following cannot be passed to cmakeFlags directly due
+    # to variable substitution
+    cmakeFlags="$cmakeFlags -DYDB_INSTALL_DIR:STRING=$out/dist"
```
Why not using cmakeFlags outside this?

ztmr replied:
How do I reference `$out/dist` from `cmakeFlags`?

AndersonTorres replied:
Just write 
```nix
  cmakeFlags = [
    (lib.cmakeOptionType "filepath" "YDB_INSTALL_DIR" "${placeholder "out'}/dist")
  ];
```

ztmr replied:
Thanks, I was not aware of this, changed.

AndersonTorres commented on pkgs/development/yottadb/default.nix:
```diff
@@ -0,0 +1,225 @@
+    util-linux
+  ];
+
+  # The version number is generated using git-describe
```
In Nixpkgs we patch such uses of git commands, substituting them by fixed strings.

ztmr replied:
I have patched the CMake mechanisms not to use Git at all and just bake in the commit reference directly into the code.

AndersonTorres reviewed (changes requested):
What are you trying to do?

Commit:
yottadb: use ASAN while running tests

Commit:
yottadb: remove patch no longer needed for new versions of YDB

Commit:
yottadb: migrate to pkgs/by-name structure

Commit:
yottadb: bump version to 2.02 and hardcode the git versions

* move away from a r1.38 fork, r2.02 has the NixOS specific patch merged
  in the official upstream
* move `systemd.tmofiles.rules` from `programs.yottadb` to the package
  itself, so that the package (and its different versions) can manage
  its own tmpfile rules
* fix ASAN integration
* patch the CMake dependency on Git describe, so that we don't need to
  keep the `.git` and do `git reset --hard` and other things
* fix the issue with CMake trying to access `/etc/os-release` (which
  not used at the end, AFAIK)
* use `substitute` instead of plain `sed`
* remove ARM platforms that we could not test on (yet)

Commit:
yottadb: reformat using nixfmt as advised by the CI

Commit:
yottadb: scratch multiversion support for now

Commit:
yottadb: nixfmt reformat

ztmr commented:
@ajs124 @AndersonTorres I'm back on this, please check again.

Note that the GH CI checks forced me to reformat the entire list of maintainers, which is insane. What's going on here?

ztmr committed:
yottadb: add mupip wrapper

ztmr commented:
@wegank who can help me getting this merged? It is painful to keep this in sync with the `master` all the time while hoping someone would actually review/approve before the `maintainer-list`, `all-tests` or `all-packages` diverge once again (all the time). Just ping me and we'll get this done.
